tools/check_reordering.py

```diff
@@ -0,0 +1,167 @@
+            continue
+
+        if symbolInfo.diff == 0:
+            symbolGood = colorama.Fore.GREEN + "GOOD" + colorama.Fore.RESET
+            if not buildFile in comparisonInfo.badFiles and not expectedFile in comparisonInfo.badFiles:
+                if not buildFile in comparisonInfo.badFiles and not expectedFile in comparisonInfo.badFiles:
+                    if not printAll:
+                        continue
+
+        if buildFile != expectedFile:
+            symbolGood += colorama.Fore.CYAN + " MOVED" + colorama.Fore.RESET
+        print(f"{symbolInfo.symbol.name},{symbolInfo.buildAddress:X},{buildFileName},{symbolInfo.expectedAddress:X},{expectedFileName},{symbolInfo.diff:X},{symbolGood}")
+
+
+def printFileComparison(comparisonInfo: mapfile_parser.MapsComparisonInfo, fun_allowed: bool = True):
+    eprint("")
+
+    if len(comparisonInfo.badFiles) != 0:
+        print(colorama.Fore.RED + "  BAD" + colorama.Style.RESET_ALL)
+
+        for file in comparisonInfo.badFiles:
+            eprint(f"Symbol reordering in {mapPathToSource(file.filepath)}")
+        eprint("")
+
+        if fun_allowed:
+            eprint(colorama.Fore.LIGHTWHITE_EX +
+            "  Symbols are REORDERED!!\n"
+            "  Oh! MY GOD!!" 
+            + colorama.Style.RESET_ALL)
+            eprint("")
+
+    if len(comparisonInfo.missingFiles) != 0:
+        print(colorama.Fore.YELLOW + "  MISSING" + colorama.Style.RESET_ALL)
+
+        for file in comparisonInfo.missingFiles:
+            eprint(f"Symbols missing from {mapPathToSource(file.filepath)}")
+        eprint("")
+
+        if fun_allowed:
+            eprint(colorama.Fore.LIGHTWHITE_EX + "  Error, should (not) be in here " + colorama.Style.RESET_ALL)
+            eprint("")
+
+        eprint("Some files appear to be missing symbols. Have they been renamed or declared as static? You may need to remake 'expected'")
+
+
+def main():
+    description = "Check that globally visible symbols has not been reordered."
+    epilog = """\
+    N.B. Since this script reads the map files, it can only see globally visible symbols; in-function static symbols must be examined with other tools.
+    """
+
+    parser = argparse.ArgumentParser(description=description, epilog=epilog, formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument("-a", "--print-all", help="Print all symbols of the section, not just non-matching.", action="store_true")
+    parser.add_argument("-n", "--no-fun-allowed", help="Remove amusing messages.", action="store_true")
+    parser.add_argument("-r", "--no-reverse-check", help="Disable looking for symbols on the expected map that are missing on the built map file.", action="store_true")
+    parser.add_argument("-s", "--section", help="Specify which section should be checked for reordered symbols. Use * to check all sections. Defaults to .bss", default=".bss")
```

Comment:
I wonder if we should have some kind of check for section names? I could easily see someone pass `bss` and think `bss` is fine when they needed to pass `.bss` instead.

Reply:
could move `mapfile_parser.MapFile()` and `readMapFile` out of `compareMapFiles` and into `main`,
then do something like
```py
all_section_types = {f.sectionType for seg in mapFile for f in seg}
if section not in all_section_types:
    approx_section_types = [section_type for section_type in all_section_types if section in section_type]
    if len(approx_section_types) == 1:
        print("Unknown section", section, "using", approx_section_types[0], "instead")
        section = approx_section_types[0]
    elif not approx_section_types:
        print("Unknown section", section)
        exit(1)
    else:
        print("Unknown section", section, "did you mean", approx_section_types)
        exit(1)
```

Reply:
I went with a hardcoded choices list on the argument instead